### PR TITLE
util, types: don't let SPM be affected by charset

### PR DESCRIPTION
**Pull request by rebelice**
### What problem does this PR solve?

Issue Number: close #xxx 

Different default charsets in different drivers(or versions) make SQL bind cannot work.
So we don't let SPM be affected by charset.

We add a flag `RestoreStringWithoutCharset` to control this.

Need to merge [parser pr](https://github.com/pingcap/parser/pull/1184) first

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 

- don't let SPM be affected by charset

### CODE REVIEW

**Review by eurekaka — COMMENTED** [review]
Please add test for this PR.

**Review by eurekaka — COMMENTED** [review]
We need a upgrade function as well.

File: session/bootstrap.go [hunk 1]
```diff
@@ -1319,8 +1322,8 @@ type bindInfo struct {
 	source     string
 }
 
-func upgradeToVer61(s Session, ver int64) {
-	if ver >= version61 {
+func upgradeToVer66(s Session, ver int64) {
+	if ver >= version66 {
```
Comment (eurekaka) [inline session/bootstrap.go:1326]: version in https://github.com/pingcap/tidb/pull/23161/files#diff-87683a21082ba99b8f8ee16504246f92fc7a8e417374ee868d00dc6bc773b015L1359 and https://github.com/pingcap/tidb/pull/23161/files#diff-87683a21082ba99b8f8ee16504246f92fc7a8e417374ee868d00dc6bc773b015R1350 should be updated as well.

File: util/parser/ast.go [hunk 2]
```diff
@@ -55,8 +55,8 @@ func RestoreWithDefaultDB(node ast.StmtNode, defaultDB string) string {
 	// Three flags for restore with default DB:
 	// 1. RestoreStringSingleQuotes specifies to use single quotes to surround the string;
 	// 2. RestoreSpacesAroundBinaryOperation specifies to add space around binary operation;
-	// 3. RestoreStringWithoutDefaultCharset specifies to not print default charset before string;
-	ctx := format.NewRestoreCtx(format.RestoreStringSingleQuotes|format.RestoreSpacesAroundBinaryOperation|format.RestoreStringWithoutDefaultCharset, &sb)
+	// 3. RestoreStringWithoutCharset specifies to not print charset before string;
+	ctx := format.NewRestoreCtx(format.RestoreStringSingleQuotes|format.RestoreSpacesAroundBinaryOperation|format.RestoreStringWithoutCharset, &sb)
```
Comment (eurekaka) [inline util/parser/ast.go:59]: Please add comment in the header of this function to inform that it is customized for SQL bind usage.

**Comment by eurekaka** [comment]
Screen Shot 2021-03-12 at 4 05 54 PM

**Review by qw4990 — APPROVED** [review]

**Review by eurekaka — COMMENTED** [review]
/LGTM

**Review by eurekaka — APPROVED** [review]
/LGTM

File: session/bootstrap_test.go [hunk 3]
```diff
@@ -545,9 +553,10 @@ func (s *testBootstrapSuite) TestUpdateBindInfo(c *C) {
 			bindCase.bindText,
 			bindCase.db,
 		)
+		fmt.Println("[sql]", sql)
```
Comment (XuHuaiyu) [inline session/bootstrap_test.go:556]: useless ?

File: session/bootstrap.go [hunk 4]
```diff
@@ -543,12 +544,13 @@ var (
 		// We will redo upgradeToVer58 in upgradeToVer64, it is skipped here.
 		upgradeToVer59,
 		upgradeToVer60,
-		upgradeToVer61,
+		// We will redo upgradeToVer61 in upgradeToVer6, it is skipped here.
```
Comment (XuHuaiyu) [inline session/bootstrap.go:547]: upgradeToVer6 ?

**Pull request merged**